coverage/control.py

```diff
@@ -862,24 +863,49 @@ def xml_report(
             ignore_errors=ignore_errors, report_omit=omit, report_include=include,
             xml_output=outfile, report_contexts=contexts,
             )
+        return self._render_report(self.config.xml_output, XmlReporter(self), morfs)
+
+    def json_report(
+        self, morfs=None, outfile=None, ignore_errors=None,
+        omit=None, include=None, contexts=None, pretty_print=None,
+        show_contexts=None
+    ):
+        """Generate an JSON report of coverage results.
+
+        Each module in `morfs` is included in the report.  `outfile` is the
+        path to write the file to, "-" will write to stdout.
+
+        See :meth:`report` for other arguments.
+
+        Returns a float, the total percentage covered.
+
+        """
+        self.config.from_args(
+            ignore_errors=ignore_errors, report_omit=omit, report_include=include,
+            json_output=outfile, report_contexts=contexts, json_pretty_print=pretty_print,
+            json_show_contexts=show_contexts
+        )
+        return self._render_report(self.config.json_output, JsonReporter(self), morfs)
+
+    @staticmethod
+    def _render_report(output_path, reporter, morfs):
         file_to_close = None
         delete_file = False
-        if self.config.xml_output:
-            if self.config.xml_output == '-':
+        if output_path:
+            if output_path == '-':
                 outfile = sys.stdout
             else:
                 # Ensure that the output directory is created; done here
                 # because this report pre-opens the output file.
                 # HTMLReport does this using the Report plumbing because
                 # its task is more complex, being multiple files.
-                ensure_dir_for_file(self.config.xml_output)
+                ensure_dir_for_file(output_path)
                 open_kwargs = {}
                 if env.PY3:
                     open_kwargs['encoding'] = 'utf8'
-                outfile = open(self.config.xml_output, "w", **open_kwargs)
+                outfile = open(output_path, "w", **open_kwargs)
                 file_to_close = outfile
         try:
-            reporter = XmlReporter(self)
```

Comment:
This moves the construction of XmlReporter out of the try/except.  I hope there isn't an edge case that try/except was handling for us.  The XmlReporter constructor doesn't do much, but it doesn't do nothing.

Reply:
ill take a look

Reply:
Yeah, this change is dangerous. the call to `coverage.get_data()` can throw the exception

I kinda feel like I should just move that out of the constructor. Not just because it makes me not have to rethink this part of the code but under the belief that constructors should exception/side effect free. I also dont think `self.data` is actually doing too much

Open to other thoughts as well of course.

Reply:
I went ahead and moved the dangerous stuff out of the constructor and into the report call